### PR TITLE
Refactor `deconstructURL ` and `scheme` parsing

### PR DESCRIPTION
**Pull request by dnadoba**
### Motivation
We want to reuse our URL parsing, validation and deconstruction logic for the upcoming new `HTTPClientRequest` from the async/await proposal.

### Changes
- Introduce a new `DeconstructedURL` struct which hold a `scheme`, `connectionTarget` and `uri`.
- Introduce a new `DeconstructedURL.Scheme` enum which replaced the previous `String` in `HTTPClient.Request` and `ConnectionPool.Scheme` enum
- move `deconstructURL` logic to an initialiser of  `DeconstructedURL`
- use `DeconstructedURL` instead of separate `scheme`, `connectionTarget`, and `uri` properties in `HTTPClient.Request`
- remove `HTTPClient.Request.Kind` because it is almost a replica of the `DeconstructedURL.Scheme` enum where the only difference is that `http` and `https` are represented as one case called `host`. As we now represent `scheme` as an enum instead of a `String` we can safely replace it with an exhaustive switch over `scheme` instead.
- refactor tests to use `scheme` instead of `kind`
- remove unused `config(overriding:)` from `ConnectionPool.Key`

### CODE REVIEW

**Comment by dnadoba** [comment]
@swift-server-bot test this please

File: Sources/AsyncHTTPClient/DeconstructedURL.swift [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
```
Comment (fabianfett) [inline Sources/AsyncHTTPClient/DeconstructedURL.swift:15]: NIT: Only import what is actually needed.
```suggestion
import struct Foundation.URL
```

File: Sources/AsyncHTTPClient/DeconstructedURL.swift [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                scheme: scheme,
+                connectionTarget: .unixSocket(path: socketPath),
+                uri: uri
+            )
+        }
+    }
+}
+
+extension DeconstructedURL.Scheme {
+    var useTLS: Bool {
```
Comment (fabianfett) [inline Sources/AsyncHTTPClient/DeconstructedURL.swift:87]: He, she, it das s muss mit ;)

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 3]
```diff
@@ -277,25 +209,28 @@ extension HTTPClient {
...
+            switch self.deconstructedURL.connectionTarget {
             case .ipAddress(_, let address): return address.port!
             case .domain(_, let port): return port
-            case .unixSocket: return Request.useTLS(self.scheme) ? 443 : 80
+            case .unixSocket: return self.deconstructedURL.scheme.defaultPort
             }
         }
 
+        /// Whether request will be executed using secure socket.
+        public var useTLS: Bool { self.deconstructedURL.scheme.useTLS }
```
Comment (fabianfett) [inline Sources/AsyncHTTPClient/HTTPHandler.swift:220]: I think `uses` is better.

Comment (fabianfett) [reply Sources/AsyncHTTPClient/HTTPHandler.swift:220]: Also, why do we add this public API.

File: Sources/AsyncHTTPClient/DeconstructedURL.swift [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+import Foundation
+
+struct DeconstructedURL {
+    enum Scheme: String {
+        case http
+        case https
+        case unix
+        case httpUnix = "http+unix"
+        case httpsUnix = "https+unix"
+    }
```
Comment (fabianfett) [inline Sources/AsyncHTTPClient/DeconstructedURL.swift:24]: I'm not sure this should be namespaced in DeconstructedURL. Would you mind moving this out?

**Review by fabianfett — COMMENTED** [review]
All in all looks good. Few changes wanted.

File: Sources/AsyncHTTPClient/HTTPHandler.swift (hunk 3 above)
Comment (dnadoba) [reply Sources/AsyncHTTPClient/HTTPHandler.swift:220]: It’s already public api and we can’t change the naming

Comment (fabianfett) [reply Sources/AsyncHTTPClient/HTTPHandler.swift:220]: Oh I see. Makes sense!

File: Sources/AsyncHTTPClient/DeconstructedURL.swift (hunk 2 above)
Comment (dnadoba) [reply Sources/AsyncHTTPClient/DeconstructedURL.swift:87]: We will diverge from the public API on HTTPClient.Request but I’m okay with that

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 5]
```diff
@@ -748,7 +748,7 @@ internal struct RedirectHandler<ResponseType> {
     }
 }
 
-extension DeconstructedURL.Scheme {
+extension Scheme {
```
Comment (fabianfett) [inline Sources/AsyncHTTPClient/HTTPHandler.swift:751]: I think this should live in the Scheme file.

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 6]
```diff
@@ -92,82 +92,19 @@ extension HTTPClient {
...
-        }
-
         /// Request HTTP method, defaults to `GET`.
         public let method: HTTPMethod
         /// Remote URL.
         public let url: URL
+
+        /// Parsed, validated and deconstructed URL.
+        internal let deconstructedURL: DeconstructedURL
+
```
Comment (fabianfett) [inline Sources/AsyncHTTPClient/HTTPHandler.swift:102]: NIT: Move down to other internal var like redirect state. Mixing internal and public makes this hard to read.

**Review by fabianfett — APPROVED** [review]
Very cool, except for the tiny nits.

**Review by fabianfett — APPROVED** [review]

**Comment by fabianfett** [comment]
@swift-server-bot test this please

**Comment by karwa** [comment]
One thing that might be worth considering is if the scheme can just be dropped?

- `http` == `httpUnix` == `unix`, the only difference is the connection target (host/UDS)
- `https` == `httpsUnix`, the only difference is the connection target (host/UDS)

Because it now knows which kind of connection target it has, the only thing the scheme provides is whether to use TLS.

**Comment by dnadoba** [comment]
Sadly we expose the scheme as public API. Not as the `Scheme` enum but as a `String` on `HTTPClient.Request` and therefore need to preserve it for the time being.

**Comment by karwa** [comment]
Technically, it could get that from the URL. There's no need for ConnectionPool or most other things to know specifically which kind of scheme was used (also, `http+unix` and `unix` schemes should resolve to the same connection in the pool if they point to the same socket).

**Comment by Lukasa** [comment]
> (also, http+unix and unix schemes should resolve to the same connection in the pool if they point to the same socket).

I disagree with this: the scheme is meaningfully part of both the origin and the network key for a request, and so should use different connections.

**Comment by fabianfett** [comment]
The problem here are http semantics. From an http standpoint `unix:///some/path` and `http+unix:///some/path` go to different places. While the server _might_ be the same, we don't know for sure. And since connections are trust boundaries, we need to make the differentiation based on the scheme.

**Comment by karwa** [comment]
I'm not sure to what extent HTTP really intends to define the semantics of non-`http:`/`https:` URLs. It sounds more likely that the semantics of application-defined schemes are also application-defined.

**Comment by Lukasa** [comment]
HTTP does not define them. I agree that in principle we can define them however we want. Given that schemes are explicitly part of the connection pooling pattern in fetch, the most prudent choice is to generalise that idea to these schemes, especially as there are very few practical problems caused by doing so.

**Pull request merged**